### ISSUE TITLE
Show Form download errors in Exactly Match Server notification

### ISSUE THREAD

**Issue by seadowg** [getodk/collect#5026]
Currently, if an error occurs while downloading a Form, rather than while fetching the Form list, the user will just see the "If you keep having this problem..." message as the notification content. This can make it hard to understand problems that lie within the Form themselves (invalid submissions, parse errors etc) despite us showing error messages for these kinds of issues when doing manual Form downloads. We should "bubble up" the Form download and show the same set of error messages we show for manual downloads (defined by `FormDownloadExceptionMapper`). The different errors to handle are:

* Server not returning hash Form in form list (`FormDownloadException.FormWithNoHash`)
* Collect can't parse the form once it's downloaded (`FormDownloadException.FormParsingError`) - this should show the original exception message so that form designers can tell what's gone wrong
* An error occurs writing the form to disk (`FormDownloadException.DiskError`)
* Form defines an incorrect/invalid submission URI (`FormDownloadException.InvalidSubmission`)
* An error occurs communicating with the server in one of the Form download steps (`FormDownloadException.FormSourceError`)

# Acceptance

- [ ] Given I've setup Collect with a server with at least one form
And I'm using Exactly Match Server
And something will cause the form download to fail
When the sync with the server occurs (manually or in the background automatically)
Then I see a notification that there has been an error
When I click on "Show details" on the notification
Then I see the error for the form that failed

- [ ] Given I've setup Collect with a server with more than one form
And I'm using Exactly Match Server
And some forms will succeed in downloading and others will fail
When the sync with the server occurs (manually or in the background automatically)
Then I see a notification that there has been an error
When I click on "Show details" on the notification
Then I see errors for the forms that failed
When I navigate to Fill Blank Form
Then I see the forms that downloaded successfully

- [ ] Given I've  setup Collect with a server with at least one form
And I'm using Exactly Match Server
And there is a form on device that's not on the server
And something will cause the form download to fail
When the sync with the server occurs (manually or in the background automatically)
Then I see a notification that there has been an error
When I navigate to Fill Blank Form
Then  the form that is not on the server is no longer there

**Comment by lognaturel** [comment]
> increasing the granularity of the messages we show in the notification is a good first step

I'm not really convinced that helps a lot. Currently, after a few repeat failures we get the idea that it's a `FormParsingError` but without knowing what that error is we're stuck. Could we do something like always show the details in the notification if it's a `FormParsingError`? I think for all the other the error type is sufficient. What you describe with the show details and the replacing is ok as well but it will always be necessary to get details in the case of `FormParsingError` so I don't see a huge benefit (unless it's much simpler to implement for some reason but my instinct is that always showing the details would be easier).

**Comment by seadowg** [comment]
@lognaturel by "details" do you mean the underlying exception (caught in `ServerFormDownloader`) message that caused us to throw a `FormParsingError`? As part of this, maybe we could alter the message for `FormParsingError` so it's:

`This form cannot be processed: <originalException.getMessage()>. If you keep having this problem, report it to the person who asked you to collect data.`

**Comment by lognaturel** [comment]
It has felt very bad when I've run into a form design issue that caused form update/download to fail but I can't find a likely-to-happen way to reproduce! If you want to see what it feels like, the following QR code has a form I deliberately mangled the XML for:
<img width="191" alt="Screen Shot 2022-02-25 at 3 27 20 PM" src="https://user-images.githubusercontent.com/967540/155816487-84ef8f73-7e56-4d0a-a928-e47283fe5309.png">

Maybe it's because it used to happen when an external secondary instance was missing and now we've addressed that? I'm really not sure. 

Given my lack of repro and the fact that we haven't actually had user complaints, I'm ok with de-prioritizing this after all. What you suggest [above](https://github.com/getodk/collect/issues/5026#issuecomment-1040038396) with including the original exception message sounds like it would be a good step to take when we get to it.

**Comment by grzesiek2010** [comment]
The current behavior is basically that all `FormDownloadExceptions` are converted into `FormSourceException.FetchError()` which is represented by that generic message. This should not happen.
But my first question is: do we want to keep that behavior that when downloading of one form fails the whole process is stopped? This is not consistent with auto-update where we try to download all forms and return a map of results. This is an important question and decision to make because if we are not going to change anything probably we can display the whole message/exception in a notification without using `ErrorActivity` since it will always be just one element otherwise we need to display the list of errors.

**Comment by grzesiek2010** [comment]
@seadowg we were talking about this issue during the meeting today and we decided to ask you if you remember why it works like that (the question above) - the process of downloading forms is stopped if any exceptions takes place what is not consistent with auto-update where we try to download all forms and collect all exceptions. Do you remember if not we probably should change it?

**Comment by seadowg** [comment]
> we were talking about this issue during the meeting today and we decided to ask you if you remember why it works like that (the question above)

I think it was just that it was that "fail fast" was the simplest way to go. If we switch to a different strategy and update as much as we can, we're going to need to decide what form update/download failures mean for deleting forms not on the list. Right now we delete all the forms and the attempt updates so could end up with an empty form list for example.

I'm not really sure whether I think that a "fail fast" or a "do as much as we can" approach is better here, but the latter certainly feels more complex to me, so we should make sure we have a good reason to go with it if we do. I don't think we should aim for consistency with auto update here as that is a fundamentally different experience.

**Comment by grzesiek2010** [comment]
> decide what form update/download failures mean for deleting forms not on the list

What do you mean by that? Could you elaborate?

**Comment by lognaturel** [comment]
> we're going to need to decide what form update/download failures mean for deleting forms not on the list

Ah, yes, I hadn't considered/remembered those were coupled somehow. It feels to me like it would be the most useful to separate the deletes from the updates and attempt all updates. That is, the form list tells us what form defs need to be deleted without attempting any form updates at all. Then all updates can be attempted and any failures reported. Maybe there's an implementation reason why it wasn't done that way but that's what I would expect as a user.

**EDIT: I missed some things and have updated my checks accordingly**

> we delete all the forms and the attempt updates so could end up with an empty form list for example

Some checks on how things work currently:
- I have 7 forms on my device. My project leader updates two forms, one of them in a broken way, and keeps all 5 others the same. Let's say the two updated forms are first in the form list with the broken one coming second. The first one will update, the second one will fail, and the other 5 that should have been kept will be deleted?
- I have 3 forms on my device. Two of them are updated in a way that needs to be consistent (e.g. they're both in the same workflow). The third is unrelated and gets updated in a broken way (or the network fails during its download? Is that possible?). Let's say one related form is first, then the unrelated form, then the second related form. The first form will have a successful update. Then the second one will fail meaning the third won't get updated and the two related forms will be out of sync. Or will it be deleted? I don't think that's what happens?

**Comment by grzesiek2010** [comment]
> and the other 5 that should have been kept will be deleted?

no, we perform deletion first but we delete only those forms that are on our device but NOT on the list of forms from the server.

**Comment by lognaturel** [comment]
> we perform deletion first but we delete only those forms that are on our device but NOT on the list of forms from the server.

Oh, good, excellent. That's what I want and how I thought I remembered it.

So here's a variation on what I wrote above for the first scenario that I think is worse than if we tried every download. I think the second scenario stands.
- I have 2 forms on the device. Those two forms are deleted on the server and two new forms are introduced. The first one is broken in some way. The second is valid. Currently the first form will fail, the update will fail fast, and we'll have an empty form list.

**Comment by grzesiek2010** [comment]
> I have 2 forms on the device. Those two forms are deleted on the server and two new forms are introduced. The first one is broken in some way. The second is valid. Currently the first form will fail, the update will fail fast, and we'll have an empty form list.

That seems correct.

**Comment by seadowg** [comment]
> I have 3 forms on my device. Two of them are updated in a way that needs to be consistent (e.g. they're both in the same workflow). The third is unrelated and gets updated in a broken way (or the network fails during its download? Is that possible?). Let's say one related form is first, then the unrelated form, then the second related form. The first form will have a successful update. Then the second one will fail meaning the third won't get updated and the two related forms will be out of sync. Or will it be deleted? I don't think that's what happens?

The two forms will be out of sync. As @grzesiek2010 stated, we only delete the forms if they're not on the form list. This example makes me lean towards a "do as much as we can" approach as right now the forms would be out of sync anyway without an error whereas if we tried downloading everything (and reported results on a form by form basis) it would be clearer what was out of sync. 

> I have 2 forms on the device. Those two forms are deleted on the server and two new forms are introduced. The first one is broken in some way. The second is valid. Currently the first form will fail, the update will fail fast, and we'll have an empty form list.

Correct, but I'm not sure that an empty form list is **worse** than two obsolete forms here?

Another alternative approach (to fail fast or download as much as we can) that is much more work would be to perform the sync in a staging area and only apply changes (deleting, updating etc) once we know we've succeeded.

**Comment by lognaturel** [comment]
> if we tried downloading everything (and reported results on a form by form basis) it would be clearer what was out of sync.

Yes, I agree with this.

> I'm not sure that an empty form list is worse than two obsolete forms here?

I think that if we proceeded after a failure we would have one obsolete and one updated. I think the form status for the first one would be "failed at timestamp", right? In a perfect world we'd display a ⚠️ or something and have inline refresh but I think what we have is good enough. 

> Another alternative approach

I don't think we need to consider this now but we may need to in an entity-based world where forms are much more connected.

**Comment by seadowg** [comment]
OK it does sound like it would be potentially better to download as many things as we can rather than failing fast. However, as far as I'm aware we're not hearing of anyone experiencing problems in the field from this and I think changing the behaviour of match exactly to fail fast would be an additional step to this issue (not a different direction). Given that, I'd vote that we just go with this issue as described (failing fast and showing more granular errors in the notification) and think about moving away from fail fast later (in addition to adding a "Show details" button as described in the issue notes).

**Comment by grzesiek2010** [comment]
If by later you mean really later like in 6 months or more then ok otherwise implementing changes in the same area in a short period of time twice would mean not only spending more time on the implementation but also on testing.

**Comment by seadowg** [comment]
> If by later you mean really later like in 6 months or more then ok otherwise implementing changes in the same area in a short period of time twice would mean not only spending more time on the implementation but also on testing.

Yup agreed. Definitely mean "later" as in "not a priority for this release". @lognaturel good for you to make the call here on whether we want to prioritize moving away from fail fast or not. I think it's worth considering, but personally I'm leaning away from it for this release.

**Comment by lognaturel** [comment]
Let's remove from this release and rework the issue so it addresses attempting every download AND showing messages for any that fail.

**Comment by grzesiek2010** [comment]
I'm adding this issue to v2022.3 @seadowg could you update the description?

**Comment by seadowg** [comment]
@grzesiek2010 I've added acceptance that hopefully covers the cases we've talked about.

**Comment by grzesiek2010** [comment]
Thanks!

**Comment by lognaturel** [comment]
Here's a case I just ran into. Recording for when we pick this back up:
- upload a form with an external select (`select_one_from_file foo`)
- attach a secondary instance without a `name` column

The first time I downloaded, the download worked. When I opened the form, I got an error message so I knew what the problem was. I [attempted to fix it](https://docs.getodk.org/form-question-types/#customizing-the-label-and-value) but made a typo. When I tried to update, it failed but I couldn't figure out why until I did an `adb logcat`.

**Comment by lognaturel** [comment]
The user who ran into [the problem with UTF-8 byte order marks](https://github.com/getodk/javarosa/issues/687) also had failures like what I described just above. Strangely, the behavior wasn't consistent. Sometimes he'd get a download error, sometimes he'd get successful download and failure to open. As part of this, let's be on the lookout for a race condition of some sort.

**Comment by lognaturel** [comment]
We're going to be asking users to edit form XML for the first preview release of entities. They're much more likely to make mistakes in that context. For example, I forgot to add the `entities:entities-version` attribute and got the dreaded generic failure message. I think this is important to release with the entities preview.

**Comment by lognaturel** [comment]
> I forgot to add the entities:entities-version attribute and got the dreaded generic failure message.

I was thinking more about this and realized that it might make more sense to make that attribute optional and fall back to the original entities spec version (2022.1.0). 

I mentioned this to @seadowg and he expressed surprise that this failure was happening on download. Now that I think about it more, I'm surprised to -- I thought we were doing a specialized pass to pull out the form title and that's it. And in fact, why are we doing any parsing on download, can't we use the title from the manifest?

@seadowg also said something broader that shifted my thinking on the importance of this issue: these form syntax issues are things we should be catching earlier in Central or pyxform. What we've done is push them to "runtime". And even then, it seems like if we are pushing them to runtime they should show up at form launch time, not at form download time.

I think there are a few things to look into:
- @lognaturel will think about modifying the `entities:entities-version` spec so that clients can have consistent behavior if it's omitted
- @seadowg/@grzesiek2010 will look into why this failure is happening when it is. Is there really a good reason to have a full parse pass?

Does it make sense for @grzesiek2010 to hold off on this issue? And/or look into these questions we have? If we could push all form parse errors to form launch that would still feel late but at least it would be at a point where a user can actually make sense of them.

**Comment by seadowg** [comment]
> @seadowg/@grzesiek2010 will look into why this failure is happening when it is. Is there really a good reason to have a full parse pass?

I can have a quick look at that this morning. 

> Does it make sense for @grzesiek2010 to hold off on this issue? And/or look into these questions we have? If we could push all form parse errors to form launch that would still feel late but at least it would be at a point where a user can actually make sense of them.

I'd vote for this. Unless you've already started @grzesiek2010 I'll unassign you and move this to the next milestone?

**Comment by grzesiek2010** [comment]
> I'd vote for this. Unless you've already started @grzesiek2010 I'll unassign you and move this to the next milestone?

I haven't. I was thinking about starting on Monday so yeah we can move it to the next milestone.

**Comment by seadowg** [comment]
I did a quick dig into why we end up parsing the form at download time. The easiest way to see what we parse out is to look at [`ServerFormDownloader#saveNewForm`](https://github.com/getodk/collect/blob/ffcb17ff5ca59c10b5643311e0164a06c79767d5/collect_app/src/main/java/org/odk/collect/android/formmanagement/ServerFormDownloader.java#L213) (`formInfo` is a map of details we parse out of the XForm):

```java
private Form saveNewForm(Map<String, String> formInfo, File formFile, String mediaPath) {
    Form form = new Form.Builder()
        .formFilePath(formFile.getAbsolutePath())
        .formMediaPath(mediaPath)
        .displayName(formInfo.get(FileUtils.TITLE))
        .version(formInfo.get(FileUtils.VERSION))
        .formId(formInfo.get(FileUtils.FORMID))
        .submissionUri(formInfo.get(FileUtils.SUBMISSIONURI))
        .base64RSAPublicKey(formInfo.get(FileUtils.BASE64_RSA_PUBLIC_KEY))
        .autoDelete(formInfo.get(FileUtils.AUTO_DELETE))
        .autoSend(formInfo.get(FileUtils.AUTO_SEND))
        .geometryXpath(formInfo.get(FileUtils.GEOMETRY_XPATH))
        .build();

    return formsRepository.save(form);
}
```

We could definitely get some of these details from the manifest (which moves a lot more trust there), but there are a bunch we can't (like submission URI or the encryption public key). Some of the details (like the submission URI) are validated at download as well, which means that forms with invalid details don't get downloaded. To me, it does feel like we should move this parse into the "first load" of the form as it feels like an easier (and more appropriate) place to present problems. That is quite a big rework though and not something I think we should attempt for this release.

I'm thinking that if we're going to relax the entities spec to eliminate the mistake we're really worried about (omitting the `entities-version` attribute), we can safely punt on this again and look at reworking our initial form load to handle these errors (and remove any parsing from the download) in the next major release. This would also have the advantage of making total download time for batches of forms faster (there would be no parsing between network requests) which would help people on unstable connections. Are you happy with that @lognaturel?

**Comment by lognaturel** [comment]
> punt on this again and look at reworking our initial form load to handle these errors (and remove any parsing from the download)

Yes, that seems ok. Probably better to close this issue and open another one for the new work but for now I've moved it to 2023.1 so we don't forget about it.

https://forum.getodk.org/t/download-failed-odk-app/39548 is another instance of a similar issue. It's not clear to me whether it's using Match Exactly or Get Blank Form but it happens when a CSV attachment is poorly-formed.

**Comment by lognaturel** [comment]
`UnrecognizedEntityVersionException` gets thrown when an old version of Collect is used with a form that uses a new version of the entities spec. For example, if I use a form with updates (entities spec version v2023.1.0) in Collect v2023.3.0, I get a mysterious error on match exactly sync.

Maybe a good first step could be to track analytics for this and see how often users are getting mysterious failures?

**Comment by lognaturel** [comment]
A user reported getting the generic "If you keep having this problem..." message. They were originally using match exactly. I asked them to switch to manual download and that also showed the generic message. Central was receiving requests for formList and the error was given on a form-by-form basis.

We ended up debugging Collect to get details and found out they were getting a `SSLException` because of a self-signed cert.

They found their way to [this forum post](https://forum.getodk.org/t/testing-odk-collect-with-a-custom-server-and-self-signed-certs-and-authentication/25571/3?u=ln) and using ngrok addressed their need.

I'm not quite sure whether this should be spun out to a different issue because it also affects manual download. Ideally we could provide raw error details in all contexts. Previously manual downloads did show those in the summary dialog.

**Comment by seadowg** [comment]
Where does the `SSLException` actually get thrown? 

My feeling is that the spirit of this issue is more around getting typical Collect users better information and instructions on who to report them to, rather helping people debug their setup. I'm wondering if we could at providing a basic HTTP log for the latter.

**Comment by lognaturel** [comment]
> Where does the SSLException actually get thrown?

I didn't check and I don't think it's important enough to dig into again now.

> typical Collect users

I think it's sometimes a bit of an unclear line to draw because I think this would also be the behavior if a self-hoster's certificate expired. But I largely agree.

And big YES to a debug log. The challenge will be knowing how much detail to include. Let's keep these notes here for now and keep thinking about how/when we want to prioritize both this and more activity logging.

**Comment by lognaturel** [comment]
Two more cases where just being able to see the error message would really help: https://github.com/getodk/collect/issues/6638 and https://forum.getodk.org/t/form-with-attachments-broken-since-v2025-1-0-custom-openrosa-backend/54050 Both are probably related to network or server responsiveness. Having the exception details surfaced to the user would help a lot with troubleshooting.

Both cases are affecting real end users who are seeing download failures without enough detail to get the error fixed. The first issue was reported by an ODK Cloud user.
